### PR TITLE
Download (get) commands and CLIs

### PR DESCRIPTION
**Pull request by alexellis**
Initial templating for kubectl and faas-cli patterns to
validate if the rules can be encoded into Go templates for
download URL patterns.

Signed-off-by: Alex Ellis (OpenFaaS Ltd) <alexellis2@gmail.com>

### CODE REVIEW

**Comment by alexellis** [comment]
@Waterdrips PTAL?

File: pkg/get/get.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	return version, nil
+}
+
+func MakeTools() []Tool {
+	tools := []Tool{
+		{
+			Owner: "openfaas",
+			Repo:  "faas-cli",
+			Name:  "faas-cli",
+			BinaryTemplate: `{{ if HasPrefix .OS "ming" -}}
```
Comment (alexellis) [inline pkg/get/get.go:136]: This works fairly well with GitHub

File: pkg/get/get.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+{{- else if eq .Arch "aarch64" -}}
+{{.Name}}-arm64
+{{- end -}}`,
+		},
+		{
+			Owner:       "kubernetes",
+			Repo:        "kubernetes",
+			Name:        "kubectl",
+			Version:     "v1.18.0",
+			URLTemplate: `https://storage.googleapis.com/kubernetes-release/release/{{.Version}}/bin/{{.OS}}/{{.Arch}}/kubectl`,
```
Comment (alexellis) [inline pkg/get/get.go:153]: This doesn't work so well with building arbitrary URLs - I need to run a function to determine the Arch value i.e. `amd64` for 64-bit MacOS. I am now thinking about having a `func` instead.

Originally the idea of a template was to be able to have this structure on disk and not compiled-in, using a `func` would prevent that.

File: pkg/get/get.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	if res.StatusCode != 302 {
+		return "", fmt.Errorf("incorrect status code: %d", res.StatusCode)
+	}
+
+	loc := res.Header.Get("Location")
+	if len(loc) == 0 {
+		return "", fmt.Errorf("unable to determine release of kubeseal")
```
Comment (Waterdrips) [inline pkg/get/get.go:124]: kubeseal --> "tool"

File: pkg/get/get.go (hunk 2 above)
Comment (alexellis) [reply pkg/get/get.go:153]: Seems like you can have a more complex template i.e.

```
`{{$arch := "arm"}}
{{- if eq .Arch "x86_64" -}}
{{$arch = "amd64"}}
{{- end -}}
			https://storage.googleapis.com/kubernetes-release/release/{{.Version}}/bin/{{.OS}}/{{$arch}}/kubectl`
```

cc @LucasRoesler @Waterdrips

Comment (LucasRoesler) [reply pkg/get/get.go:153]: if you want it on disk, then i think using an gotmpl input is required. The simplest thing is to probably just pass the Tool object in as the template context

**Pull request merged**